### ISSUE TITLE
NPE using Http4K or WebSocketPublisher as server

### ISSUE THREAD

**Issue by davidwynter** [RuedigerMoeller/kontraktor#60]
I updated to 4.20 and am now getting these errors when trying to connect to a Http4K:

    I 10:24:12:481;ExampleHttpApp Dispatcher;io.undertow.websockets.jsr  UT026004: Adding annotated client endpoint class org.nustaq.kontraktor.remoting.websockets.JSR356ClientConnector$WSClientEndpoint
    java.lang.NullPointerException
	at org.nustaq.kontraktor.Promise.<init>(Promise.java:86)
	at org.nustaq.kontraktor.remoting.websockets.JSR356ClientConnector.connect(JSR356ClientConnector.java:77)
	at org.nustaq.kontraktor.remoting.base.ActorClient.connect(ActorClient.java:72)
	at org.nustaq.kontraktor.remoting.websockets.WebSocketConnectable.lambda$connect$0(WebSocketConnectable.java:69)
	at org.nustaq.kontraktor.remoting.websockets.WebSocketConnectable.connect(WebSocketConnectable.java:80)
	at org.nustaq.kontraktor.remoting.base.ConnectableActor.connect(ConnectableActor.java:50)
	at uk.co.example.InitServicesHelper.initExampleServices(InitServicesHelper.java:34)

I am still using this pattern for the server:

		// as websocket service, fast serialized
		// pass in the host IP address		
		Http4K.Build(hostname, port)
	        .resourcePath("/")
			    .buildResourcePath()
			.websocket("/example", service)
			    .serType(SerializerType.JsonNoRef)
			    .buildWebsocket()
			.build();

 and this for the client:

			ConnectableActor initiconnectable = new WebSocketConnectable(ExampleService.class,
					"ws://" + examplehostname + ":" + ExampleHttpApp.exampleport + "/example").serType(SerializerType.JsonNoRef);

			ExampleService exampleService = null;
			try {
				exampleService = (ExampleService) initiconnectable
					.connect((res, err) -> logger.error(initiconnectable + " not connected!")).await();
			} catch(AwaitException ae) {
    ...

I note the test class now uses ActorServer for the server and still uses WebSocketConnectable for the client, but with a new coding interface.

        ActorServer publisher = new WebSocketPublisher(service, "localhost", "/ws", 8081).coding(coding).publish().await();
        RemotingTestService client = (RemotingTestService)
            new WebSocketConnectable(RemotingTestService.class, "ws://localhost:8081/ws")
                .coding(coding)
                .connect()
                .await(9999999);

I do not see a test for Http4K anymore?

So can you confirm that Http4K is no longer supported?

**Comment by davidwynter** [comment]
I tried WebSocketPublisher and got the same NPE exception.

    private static ExampleService service = null;

...

    service = Actor.AsActor(ExampleService.class);
    ActorServer publisher = new WebSocketPublisher(service, hostname, "/example", port)
				.coding(coding).publish().await();

**Comment by RuedigerMoeller** [comment]
oops, no this is accidental .. will have a look into

**Comment by RuedigerMoeller** [comment]
seem like I got bitten by Overloading ..

I added this constructor to Promise class:
```java
    /**
     * Promise constructor matching ES6-style Promise constructor semantics
     *
     * return new Promise( (resolve,reject) -> { if ( 1 != 0 ) resolve.accept("Hello") else reject.accept("Never") }
     *
     * @param <T>
     */
    public Promise(BiConsumer<Consumer<T>,Consumer<Object>> body) {
        body.accept( resolved -> this.resolve(resolved), rejected -> this.reject(rejected) );
    }
```

seems like new Promise(null,_) now chooses this one. Will remove the new Cosntructor (did not workout anyways as generics + type inference just did not figure out this for most real world cases ..)

**Comment by RuedigerMoeller** [comment]
there are obviously not enough unit tests and currently I am testing against my commercial work system which uses JS<=>Java via Http Longpoll and does service communication via TCP ...

**Comment by davidwynter** [comment]
Thx, 

So is WebSocketPublisher or Http4K preferred for connection? Or have I missed something different about the behaviour of these 2 that differentiates them? (Want to have a peek at my other issue, just below this one?)

**Comment by RuedigerMoeller** [comment]
Http4k just pools webserver instances (for Http/WebSocket connections), so its used internally regardless :).  

You can use either "Publisher" or Http4k Builder .. its same

(see new Testcase https://github.com/RuedigerMoeller/kontraktor/blob/trunk/modules/kontraktor-http/src/test/java/remoting/RemotingTest.java)

**Comment by RuedigerMoeller** [comment]
fixed with 4.22 (already uploaded, takes some hours until visible on maven central).

Added Testcase (http, websocket, tcp, tcpnio) https://github.com/RuedigerMoeller/kontraktor/blob/trunk/modules/kontraktor-http/src/test/java/remoting/RemotingTest.java
